Update C++ documentations (#21925)
DataDog/documentation: content/en/tracing/trace_collection/automatic_instrumentation/dd_libraries/cpp.md

type: multi-code-lang
code_lang_weight: 50
further_reading:

> - link: "https://github.com/DataDog/dd-opentracing-cpp"

Reviewer: Should we add a note to the top of this repo letting folks know to use dd-trace-cpp instead?

Author: Good idea. First step is I add the note at the top and soon we should deprecate the repo. What do you think?

Reviewer: Assuming you mean archive/read-only, then 100% agreed

> - link: "https://github.com/DataDog/dd-trace-cpp"

Reviewer: Do we need to notify customers about this change?

@tushar-datadog

  dd::TracerConfig config;
  config.defaults.service = "my-service";

  const auto validated_config = dd::finalize_config(config);
  if (!validated_config) {
    std::cerr << validated_config.error() << '\n';
    return 1;
  }

  dd::Tracer tracer{*validated_config};
  // Create some spans.
  {
    auto span_a = tracer.create_span();
    span_a.set_name("A");
    span_a.set_tag("tag", "123");
    auto span_b = span_a.create_child();
    span_b.set_name("B");
    span_b.set_tag("tag", "value");
  }

  return 0;
}
```

```bash
cmake -B build -DDD_TRACE_VERSION=v0.1.12 .
cmake --build build --target tracer_example -j

./build/tracer_example
DATADOG TRACER CONFIGURATION - {"collector":{"config":{"event_scheduler":{"type":"datadog::tracing::ThreadedEventScheduler" ... }}}
```

{{% /tab %}}

{{% tab "Manual" %}}

```bash
# Requires the "jq" command, which can be installed via
# the package manager:
#   - APT: `apt install jq`
#   - APK: `apk add jq`
#   - YUM: `yum install jq`
if ! command -v jq >/dev/null 2>&1; then
  >&2 echo "jq command not found. Install using the local package manager."
  exit 1
fi

# Gets the latest release version number from GitHub.
get_latest_release() {

>   curl --silent "https://api.github.com/repos/$1/releases/latest" | jq --raw-output .tag_name

Reviewer: We can eliminate `jq` and make this dependency-less via `sed`
```suggestion
  curl --silent "https://api.github.com/repos/$1/releases/latest" | grep -o '"tag_name": "[^"]*' | sed 's/"tag_name": "//;s/",//'
```

  tag: "Advanced Usage"
  text: "Advanced Usage"
---

<div class="alert alert-warning">
  <strong>Note:</strong> C++ does not provide integrations for automatic instrumentation, but it's used by Proxy tracing such as <a href="/tracing/setup/envoy/">Envoy</a> and <a href="/tracing/setup/nginx/">Nginx</a>.
</div>

## Compatibility requirements

> The C++ tracing library requires C++17 toolchain to build. For a full list of Datadog's tracing library compatiblity requirements and processor architecture support, visit the [Compatibility Requirements][3] page.

Reviewer 2: ```suggestion
The C++ tracing library requires a C++17 compiler to build. For a full list of Datadog's tracing library compatibility requirements and processor architecture support, visit the [Compatibility Requirements][3] page.
```
^ Is it more accurate to say this needs a C++17 compiler instead of toolchain?

Author: A toolchain is a set of tools (such as a compiler, linker, and assembler). A compiler transform code to binary format. A compiler on its own will no generate the static/shared library, I do think toolchain is more appropriate.

## Getting started
Before you begin, make sure you have already [installed and configured the Agent][6].

## Instrument your application

{{< tabs >}}

> {{% tab "CMake" %}}

Reviewer 2: ```suggestion
{{% tab "CMake" %}}
To integrate the `dd-trace-cpp` library into your C++ project using CMake, follow these steps:
```
^ I think we add more introduction to these code examples so it's clear for users. Something like this (please check).

````CMake
include(FetchContent)

FetchContent_Declare(
  dd-trace-cpp
  GIT_REPOSITORY https://github.com/DataDog/dd-trace-cpp
  GIT_TAG        v0.2.0
  GIT_SHALLOW    ON
  GIT_PROGRESS   ON
)

FetchContent_MakeAvailable(dd-trace-cpp)

# Add `tracer_example` target
add_executable(tracer_example tracer_example.cpp)

# Statically link against `dd-trace-cpp`
# NOTE: To dynamically link against `dd-trace-cpp` use the `dd_trace_cpp_shared` target
target_link_libraries(tracer_example dd_trace_cpp-static)
````

> 

Reviewer 2: ```suggestion
Here's an example of how to use the dd-trace-cpp library in your C++ code to create spans and traces:
```
^ I think we need to add more introduction to these code examples so it's clear for users. Something like this (please check).

```cpp
// tracer_example.cpp
#include <datadog/span_config.h>
#include <datadog/tracer.h>
#include <datadog/tracer_config.h>

#include <iostream>
#include <string>

namespace dd = datadog::tracing;

int main() {
  dd::TracerConfig config;
  config.service = "my-service";

  const auto validated_config = dd::finalize_config(config);
  if (!validated_config) {
    std::cerr << validated_config.error() << '\n';
    return 1;
  }

  dd::Tracer tracer{*validated_config};
  // Create some spans.
  {
    auto span_a = tracer.create_span();
    span_a.set_name("A");
    span_a.set_tag("tag", "123");
    auto span_b = span_a.create_child();
    span_b.set_name("B");
    span_b.set_tag("tag", "value");
  }

  return 0;
}
```

```bash
cmake -B build .
cmake --build build --target tracer_example -j

./build/tracer_example
DATADOG TRACER CONFIGURATION - {"collector":{"config":{"event_scheduler":{"type":"datadog::tracing::ThreadedEventScheduler" ... }}}
```

{{% /tab %}}

{{% tab "Manual" %}}

> 

Reviewer 2: ```suggestion
To manually download and install the dd-trace-cpp library, run the following bash script:
```
^ I think we need to add more introduction to these code examples so it's clear for users. Something like this (please check).

```bash
# Requires the "jq" command, which can be installed via
# the package manager:
#   - APT: `apt install jq`
#   - APK: `apk add jq`
#   - YUM: `yum install jq`
if ! command -v jq >/dev/null 2>&1; then
  >&2 echo "jq command not found. Install using the local package manager."
  exit 1
fi

# Gets the latest release version number from GitHub.
get_latest_release() {
  curl --silent "https://api.github.com/repos/$1/releases/latest" | jq --raw-output .tag_name
}

DD_TRACE_CPP_VERSION="$(get_latest_release DataDog/dd-trace-cpp)"

# Download and install dd-trace-cpp library.
wget https://github.com/DataDog/dd-trace-cpp/archive/${DD_TRACE_CPP_VERSION}.tar.gz -O dd-trace-cpp.tar.gz
mkdir dd-trace-cpp && tar zxvf dd-trace-cpp.tar.gz -C ./dd-trace-cpp/ --strip-components=1
cd dd-trace-cpp

# Download and install the correct version of dd-trace-cpp.
# Configure the project, build it, and install it.
cmake -B build .
cmake --build build -j
cmake --install build
```

> 

Reviewer 2: ```suggestion
After installing the `dd-trace-cpp` library, you can use it in your C++ code to create spans and traces. Here's an example:
```
^ I think we need to add more introduction to these code examples so it's clear for users. Something like this (please check).

[4]: https://app.datadoghq.com/apm/service-setup
[5]: /tracing/trace_collection/library_config/cpp/
[6]: /tracing/trace_collection/automatic_instrumentation/?tab=datadoglibraries#install-and-configure-the-agent

> [7]: TODO

Reviewer: This looks out of place 
```suggestion
```
